src/algebra/squarefree.lean

```diff
@@ -203,10 +203,41 @@ begin
   by_cases hy : y = 0,
   { simp [hy, zero_pow (nat.pos_of_ne_zero h0)] },
   refine ⟨λ h, _, λ h, h.pow h0⟩,
+  haveI := @unique_factorization_monoid.normalization_monoid R _ ⟨⟨x, 0, hx⟩⟩ _,
   rw [dvd_iff_normalized_factors_le_normalized_factors hx (pow_ne_zero n hy),
     normalized_factors_pow,
     ((squarefree_iff_nodup_normalized_factors hx).1 hsq).le_nsmul_iff_le h0] at h,
   rwa dvd_iff_normalized_factors_le_normalized_factors hx hy,
 end
 
+lemma _root_.squarefree.dvd_of_pow_dvd {x y : R} {n : ℕ}
+  (hsq : squarefree x) (h : x ∣ y ^ n) : x ∣ y :=
+if h0 : n = 0
+  then trans (by rwa [h0, pow_zero] at h) (one_dvd y)
+  else (dvd_pow_iff_dvd_of_squarefree hsq h0).1 h
+
+lemma squarefree_iff_dvd_of_pow_dvd {x : R} (h0 : x ≠ 0) :
+  squarefree x ↔ (∀ y (n : ℕ), x ∣ y ^ n → x ∣ y) :=
+⟨λ hsq y n, hsq.dvd_of_pow_dvd, λ h, begin
+  classical,
+  haveI := nontrivial_of_ne x 0 h0,
+  haveI := @unique_factorization_monoid.normalization_monoid R _ _ _,
+  obtain ⟨n, hn⟩ := (normalized_factors x).le_smul_dedup,
+  have := multiset.prod_dvd_prod_of_le hn,
+  rw multiset.prod_nsmul at this,
+  apply squarefree_of_dvd_of_squarefree (h _ n $ (normalized_factors_prod h0).symm.dvd.trans this),
+  have := λ a ha, irreducible_of_normalized_factor a (multiset.mem_dedup.1 ha),
+  rw [squarefree_iff_nodup_normalized_factors, normalized_factors_prod_eq],
+  { refine multiset.nodup.map_on (λ x hx y hy he, _) (multiset.nodup_dedup _),
+    rw multiset.mem_dedup at hx hy,
+    rwa [normalize_normalized_factor x hx, normalize_normalized_factor y hy] at he },
+  exacts [this, multiset.prod_ne_zero (λ h, (this 0 h).ne_zero rfl)],
```

Comment:
This direction should hold without the `unique_factorization_monoid` assumption.
If `r² | x`, say `x = kr²`, then `x | (kr)²`. Applying the hypothesis we have `x | kr`, which implies `rx | kr²`, i.e. `rx | x` and we're done.

Reply:
Thanks, this is brilliant! I've formalized the radical->squarefree direction for `cancel_comm_monoid_with_zero`. On the other hand I read in [this paper](https://d-nb.info/1192415523/34) (page 5) that the squarefree->radical direction holds for GCD domains, and I'll work out a proof in Lean.

Reply:
I've incorporated your proof and proved the equivalence for all gcd_monoid.